### PR TITLE
nested course subject and task

### PR DESCRIPTION
**Pull request by Hoang-Tung**

### CODE REVIEW

File: app/controllers/supervisor/courses_controller.rb [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+    else
+      render "edit"
+    end
+  end
+
+  private
+  def course_params
+    params.require(:course).permit(:name, :info, :total_level, :status,
+      course_subjects_attributes: [:id, :subject_id, :_destroy], 
+      user_courses_attributes: [:id, :user_id, :course_id, :_destroy])
```
Comment (chuanhtuan) [inline app/controllers/supervisor/courses_controller.rb]: separate user_courses to a other function, "Assign trainee"

File: app/assets/javascripts/supervisor/user_courses.coffee [hunk 2]
```diff
@@ -0,0 +1,3 @@
+# Place all the behaviors and hooks related to the matching controller here.
```
Comment (dieunb) [inline app/assets/javascripts/supervisor/user_courses.coffee]: remove this file

File: app/assets/stylesheets/supervisor/user_courses.scss [hunk 3]
```diff
@@ -0,0 +1,3 @@
+// Place all the styles related to the supervisor/UserCourses controller here.
```
Comment (dieunb) [inline app/assets/stylesheets/supervisor/user_courses.scss]: remove this file

File: app/controllers/supervisor/user_courses_controller.rb [hunk 4]
```diff
@@ -0,0 +1,17 @@
+class Supervisor::UserCoursesController < ApplicationController
+  before_action :id_course, only: [:new, :create]
+
+  def new
+  	@enroll = @course.user_courses.new
+  	@users = User.free_trainer
+    byebug
```
Comment (dieunb) [inline app/controllers/supervisor/user_courses_controller.rb]: remove

File: app/controllers/supervisor/user_courses_controller.rb [hunk 5]
```diff
@@ -0,0 +1,17 @@
+class Supervisor::UserCoursesController < ApplicationController
+  before_action :id_course, only: [:new, :create]
+
+  def new
+  	@enroll = @course.user_courses.new
```
Comment (dieunb) [inline app/controllers/supervisor/user_courses_controller.rb]: indent

File: app/controllers/supervisor/user_courses_controller.rb [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+  	@users = User.free_trainer
+    byebug
+  end
+
+  def destroy
+  end
+
+  private 
+  def id_course
+  	@course = Course.find params[:course_id]
```
Comment (dieunb) [inline app/controllers/supervisor/user_courses_controller.rb]: indent

File: app/models/user_course.rb [hunk 7]
```diff
@@ -0,0 +1,4 @@
+class UserCourse < ActiveRecord::Base
+	belongs_to :course, class_name: "Course"
```
Comment (dieunb) [inline app/models/user_course.rb]: indent

File: app/views/supervisor/courses/show.html.erb [hunk 8]
```diff
@@ -0,0 +1,17 @@
+<h1><%= @course.name %></h1>
+<h3>
+	<%= link_to "edit course", edit_supervisor_course_path(@course) %>
```
Comment (dieunb) [inline app/views/supervisor/courses/show.html.erb]: inden

File: app/models/course.rb [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+  has_many :course_subjects, dependent: :destroy
+	
+  accepts_nested_attributes_for :course_subjects, allow_destroy: true
+  accepts_nested_attributes_for :subjects, reject_if: lambda {|a| a[:name].blank?}, 
+    allow_destroy: true
+  accepts_nested_attributes_for :user_courses, allow_destroy: true
+  accepts_nested_attributes_for :users, reject_if: lambda { |a| a[:name].blank? },
+    allow_destroy: true
+	
+  validates :name, presence: true, length: { maximum: 400 },
```
Comment (dieunb) [inline app/models/course.rb]:
```
{maximum: 400}
```

File: app/models/subject.rb [hunk 10]
```diff
@@ -1,8 +1,10 @@
...
+  belongs_to :course
+	
+  has_many :activate_course_subjects, foreign_key: "subject_id", dependent: :destroy
+  has_many :courses, through: :activate_course_subjects
 
-  accepts_nested_attributes_for :tasks,
-    reject_if: {|a| a[:name].blank?}, allow_destroy: true
-  validates :name, presence: true, length: {maximum: 400},
+  accepts_nested_attributes_for :tasks, allow_destroy: true
+  validates :name, presence: true, length: { maximum: 400 },
```
Comment (dieunb) [inline app/models/subject.rb]:
```
{maximum: 400}
```

File: app/models/user.rb [hunk 11]
```diff
@@ -2,6 +2,10 @@ class User < ActiveRecord::Base
   attr_accessor :remember_token
 
   before_save {self.email = email.downcase}
+
+  has_many :user_courses, foreign_key: "user_id", dependent: :destroy
```
Comment (dieunb) [inline app/models/user.rb]: use `'` instead of `"`

File: app/views/supervisor/courses/edit.html.erb [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+      <%= f.text_field :total_level, class: "form-control" %>
+      </div>
+      <div class="form-group">
+      <%= f.label :status %>
+      <%= f.text_field :status, class: "form-control" %>
+      </div>
+    </div>
+  </div>
+  <%= f.collection_check_boxes :subject_ids, Subject.all, :id, :name do |b| %>
+        <tr>
```
Comment (dieunb) [inline app/views/supervisor/courses/edit.html.erb]: indent

File: app/views/supervisor/courses/edit.html.erb [hunk 13]
```diff
@@ -0,0 +1,41 @@
...
+          <%= b.label do %>
+            <td><%= b.text %></td>
+            <td class="selection_checkbox"><%= b.check_box %></td>
+          <% end %>
+        </tr>
+  <% end %>
+  <div class="form-group">
+      <%= f.submit "Save changed" , class: "btn btn-primary" %>
```
Comment (dieunb) [inline app/views/supervisor/courses/edit.html.erb]: indent

File: app/views/supervisor/courses/new.html.erb [hunk 14]
```diff
@@ -0,0 +1,39 @@
...
+      	<%= f.text_field :total_level, class: "form-control" %>
+      </div>
+      <div class="form-group">
+      	<%= f.label :status %>
+      	<%= f.text_field :status, class: "form-control" %>
+      </div>
+    </div>
+  </div>
+  <h3>Subject</h3>
+  <%= f.collection_check_boxes :subject_ids, Subject.all, :id, :name do |b| %>
```
Comment (dieunb) [inline app/views/supervisor/courses/new.html.erb]: indent
```
+  <%= f.collection_check_boxes :subject_ids, Subject.all, :id, :name do |b| %>
+        <tr>
+          <%= b.label do %>
+            <td><%= b.text %></td>
+            <td class="selection_checkbox"><%= b.check_box %></td>
+          <% end %>
+        </tr>
+  <% end %>
```

**Pull request merged**